packages/typescript-estree/src/convert.ts

```diff
@@ -97,12 +100,12 @@ export class Converter {
    * @param allowPattern flag to determine if patterns are allowed
    * @returns the converted ESTree node
    */
-  private converter(
-    node?: ts.Node,
+  private converter<P extends boolean>(
+    node: TSNodeConvertable,
     parent?: ts.Node,
     inTypeMode?: boolean,
-    allowPattern?: boolean,
-  ): any {
+    allowPattern?: P,
+  ): TSESTreeToTSNodeGuard<typeof node, P> {
```

Comment:
I don't think this works the way you think it will - typescript will not parameterise the passed type of `node`, meaning this type will be the same as `TSESTreeToTSNodeGuard<TSNodeConvertable, P>`

Reply:
yea i know that this type check is not working, and it seem that there is no easy way around it, for now i just wanted to validate if we are missing anything in ast, best aproach i can think of is

```ts
  private convertJSXTagName<T extends ts.JsxTagNameExpression>(
    node: T,
    parent: ts.Node,
  ): T extends ts.JsxTagNamePropertyAccess
    ? TSESTree.JSXMemberExpression
    : T extends ts.ThisExpression | ts.Identifier
    ? TSESTree.JSXIdentifier
    : TSESTree.JSXMemberExpression | TSESTree.JSXIdentifier;
  private convertJSXTagName(
    node: ts.JsxTagNameExpression,
    parent: ts.Node,
  ): TSESTree.JSXMemberExpression | TSESTree.JSXIdentifier {
```

different function but same issue

Reply:
https://www.typescriptlang.org/play?ts=4.1.5#code/JYOwLgpgTgZghgYwgAgEJwM4QHIHsAmKA3gLABQylykGYAjAFzK1SgDm5AvueaJLIhR5CdZBAAekEPgxpMOAsXJVkAa1D4mdZVRi5cTFuy48yfaPCTJhEAExjJEabPRYbyUhSrrpTWzsoAIzgoQzBWEA4ybjJyGABXEAQwYFwQaghaUQAKEEUmG1EAH2tFWwBKJjAATwAHCFwYZDzCBykZUpFkAH5kAHI6PuQquoamlpQJdtkbe16+2yGmEHiAG1WPAORgJtzFADoffGQAXjPkOnLNrxVKCf29XFP+wa2VAHp35AABMAwAWgk9WSgKgUFwUDeVHuwSgzwGfShlCgEDA8Sg6QRW04YlWWG2u3uR1O5wq11u0IOsPhiyRyE+Pz+gPEwLAoPBkJut3ujxpiK5KhRaIx-VpXJigtR6PSK3WJliZASSRSaQytFsAB4ACptJwdQrIEqzAB8yD2hCYWsqyB1Uz1M0Uonmg2GNt1zk6dh6oqWzTWG08Kh2ZqJGhJJwuV0DFOaB15EaxAqoDN+AKBEBB0A5dJhIXhryTyKlIsTKhxEDxKGD5oghzDZwjZOjFNzcITYpj9K+qeZrPZEJzcf0fLpQulPuxWzHItlq3lQA